git/index/base.py

```diff
@@ -360,20 +361,19 @@ def from_tree(cls, repo: "Repo", *treeish: Treeish, **kwargs: Any) -> "IndexFile
         # as it considers existing entries. moving it essentially clears the index.
         # Unfortunately there is no 'soft' way to do it.
         # The TemporaryFileSwap assure the original file get put back
-        if repo.git_dir:
-            index_handler = TemporaryFileSwap(join_path_native(repo.git_dir, "index"))
         try:
-            repo.git.read_tree(*arg_list, **kwargs)
-            index = cls(repo, tmp_index)
-            index.entries  # force it to read the file as we will delete the temp-file
-            del index_handler  # release as soon as possible
+            with ExitStack() as stack:
+                if repo.git_dir:
```

Comment:
I wonder why that was necessary - wouldn't `git_dir` always be set? Otherwise it wouldn't be usable, right?
If that check was unnecessary, I assume the `ExitStack` could be left out as well.

Reply:
I made some refactoring here to remove the finally block